server/src/main/java/org/elasticsearch/action/admin/cluster/stats/TransportClusterStatsAction.java

```diff
@@ -127,41 +150,19 @@ protected void newResponseAsync(
                 + "the cluster state that are too slow for a transport thread"
         );
         assert ThreadPool.assertCurrentThreadPool(ThreadPool.Names.MANAGEMENT);
-        assert task instanceof CancellableTask;
-        final CancellableTask cancellableTask = (CancellableTask) task;
-        final ClusterState state = clusterService.state();
-        final Metadata metadata = state.metadata();
-        final ClusterSnapshotStats clusterSnapshotStats = ClusterSnapshotStats.of(
-            state,
-            clusterService.threadPool().absoluteTimeInMillis()
-        );
-
-        final ListenableFuture<MappingStats> mappingStatsStep = new ListenableFuture<>();
-        final ListenableFuture<AnalysisStats> analysisStatsStep = new ListenableFuture<>();
-        mappingStatsCache.get(metadata, cancellableTask::isCancelled, mappingStatsStep);
-        analysisStatsCache.get(metadata, cancellableTask::isCancelled, analysisStatsStep);
-        mappingStatsStep.addListener(
-            listener.delegateFailureAndWrap(
-                (l, mappingStats) -> analysisStatsStep.addListener(
-                    l.delegateFailureAndWrap(
-                        (ll, analysisStats) -> ActionListener.completeWith(
-                            ll,
-                            () -> new ClusterStatsResponse(
-                                System.currentTimeMillis(),
-                                metadata.clusterUUID(),
-                                clusterService.getClusterName(),
-                                responses,
-                                failures,
-                                mappingStats,
-                                analysisStats,
-                                VersionStats.of(metadata, responses),
-                                clusterSnapshotStats
-                            )
-                        )
-                    )
-                )
+        additionalStatsListener.andThenApply(
+            additionalStats -> new ClusterStatsResponse(
+                System.currentTimeMillis(),
+                additionalStats.clusterUUID(),
+                clusterService.getClusterName(),
```

Comment:
It seems to me a bit arbitrary that `clusterUUID` here is part of `additionalStats` but cluster name is not. Both depend on the same cluster service as it seems, both should be pretty cheap to compute. It's not a real problem just not entirely clear why one is there and another is here.

Reply:
Yeah arbitrary indeed - `clusterUUID` is part of the cluster state, which we have to read up-front and I didn't want to read it again later, whereas `clusterName` is a static setting so it doesn't really matter where we get it.